mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,370 @@
+                        Value operand = it->first->getOperand(i);
+
+                        it->second->replaceUsesOfWith(
+                            it->second->getOperand(i),
+                            // entrybb->getArgument(
+                            //     std::find(ArgOps.begin(), ArgOps.end(), OperandSource) -
+                            //     ArgOps.begin()));
+                            entrybb->getArgument(FuncArg2BlockArg[operand]));
+                    }
+                }
+                else {
+                    Value operand = it->first->getOperand(i);
+
+                    it->second->replaceUsesOfWith(it->second->getOperand(i),
+                                                  entrybb->getArgument(FuncArg2BlockArg[operand]));
+                }
+            }
+        }
+
+        // Injection done, erase the original operations
+        // Note that some op in TapeOps are in NecessaryOpsForPostProcessing
+        // so remove them first
+        for (auto op : *TapeOps) {
+            auto where = std::find(NecessaryOpsForPostProcessing.begin(),
+                                   NecessaryOpsForPostProcessing.end(), op);
+            if (where != NecessaryOpsForPostProcessing.end()) {
+                NecessaryOpsForPostProcessing.erase(where);
+            }
+            op->dropAllUses();
+            op->erase();
+        }
+    }
+
+    void runOnOperation() override
+    {
+        Operation *module = getOperation();
+        OpBuilder builder(module->getContext());
+
+        // 1. Identify the functions with multiple tapes
+        // Walk through each function and count the number of devices.
+        // In frontend when tracing (jax_tracers.py/trace_quantum_function),
+        // each tape has its own quantum.device operation attached to it
+        SmallVector<func::FuncOp> MultitapePrograms;
+        module->walk([&](func::FuncOp func) {
+            if (countTapes(func) >= 2) {
+                MultitapePrograms.push_back(func);
+            }
+        });
+
+        // Do nothing and exit for classical and single-tape programs
+        if (MultitapePrograms.empty()) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+
+        // for (auto func : MultitapePrograms){
+        func::FuncOp func =
+            MultitapePrograms[0]; // temporary! TODO: process all multitape functions
+        SmallVector<SmallVector<Operation *> *> OpsEachTape(countTapes(func) + 2, nullptr);
+
+        CollectOperationsForEachTape(func, OpsEachTape);
+        //}
+
+        // 3. Get the SSA values needed by the post processing (PP)
+        // These need to be returned by the tapes
+        SmallVector<Operation *> NecessaryOpsForPostProcessing;
+
+        // Go through all the operands of all the PP ops
+        // Find the ones that are not produced in PP itself
```

Comment:
the PP opss regions' ops' operands need to be looked at as well (if not isolated from above); they can use upstream values too

Reply:
this is easily doable by a walker into the ops

Reply:
done